services/ndd_service.js

```diff
@@ -4,155 +4,106 @@
-    this._nddStoreWatchers = [];
-    this._running = new Set();
-    this._sockets = new Map();
-    this._frontend = null;
-  }
+process.on('uncaughtException', silentRpcErrors);
+process.on('unhandledRejection', silentRpcErrors);
 
-  async init(frontend) {
-    this._frontend = frontend;
-    this._nddStores = [await fsMkdtemp(path.join(os.tmpdir(), 'ndb-'))];
-
-    try {
-      const ndbDir = path.join(os.homedir(), '.ndb');
-      if (!fs.existsSync(ndbDir))
-        fs.mkdirSync(ndbDir);
-      const nddStoreDir = path.join(ndbDir, 'ndd_store');
-      if (!fs.existsSync(nddStoreDir))
-        fs.mkdirSync(nddStoreDir);
-      this._nddStores.push(nddStoreDir);
-    } catch (e) {
-    }
+const DebugState = {
+  WS_OPEN: 1,
+  WS_ERROR: 2,
+  WS_CLOSE: 3,
+  PROCESS_DISCONNECT: 4
+};
 
-    this._nddStoreWatchers = [];
-    for (const nddStore of this._nddStores) {
-      const watcher = chokidar.watch(nddStore, {
-        awaitWriteFinish: {
-          stabilityThreshold: 100,
-          pollInterval: 100
-        },
-        cwd: nddStore,
-        depth: 0
-      });
-      this._nddStoreWatchers.push(watcher);
-      watcher.on('add', id => {
-        this._running.add(id);
-        this._onAdded(nddStore, id);
-      });
-      watcher.on('unlink', id => this._running.delete(id));
-      watcher.on('error', error => 0);
-    }
-  }
+const CALL_EXIT_MESSAGE = JSON.stringify({
+  id: -1,
+  method: 'Runtime.evaluate',
+  params: { expression: 'process.exit(-1)' }
+});
 
-  nddStore() {
-    return this._nddStores[0];
-  }
+class NddService {
+  constructor(frontend) {
+    process.title = 'ndb/ndd_service';
+    this._disconnectPromise = new Promise(resolve => process.once('disconnect', () => resolve(DebugState.PROCESS_DISCONNECT)));
+    this._sockets = new Map();
 
-  async _onAdded(nddStore, id) {
-    try {
-      const info = JSON.parse(await fsReadFile(path.join(nddStore, id), 'utf8'));
-      let inspectorUrl = info.inspectorUrl;
-      delete info.inspectorUrl;
-      if (inspectorUrl.startsWith('http:')) {
-        const targetInfo = (await this._fetch(inspectorUrl))[0];
-        inspectorUrl = targetInfo.webSocketDebuggerUrl;
-      }
-      const ws = new WebSocket(inspectorUrl);
-      ws.once('open', () => {
-        this._sockets.set(id, ws);
-        this._frontend.detected(id, info);
-      });
-      ws.on('message', rawMessage => {
-        const message = JSON.parse(rawMessage);
-        message.sessionId = id;
-        this._frontend.dispatchMessage(message);
+    const localRelative = path.relative(process.cwd(), `${os.tmpdir()}/`);
+    const pipePrefix = process.platform === 'win32' ? '\\\\.\\pipe\\' : localRelative;
+    const pipeName = `node-ndb.${process.pid}.sock`;
+    this._pipe = path.join(pipePrefix, pipeName);
+    const server = net.createServer(socket => {
+      socket.on('data', async d => {
+        await new Promise(resolve => this._runSession(JSON.parse(d.toString()), frontend, resolve));
```

Comment:
`_runSession` is already async. Why do you need to wrap it?

Reply:
I tried to split it.